### PR TITLE
add functions to queries

### PR DESCRIPTION
**Pull request by nlandolfi**
Desired Query Capabilities:
- [X] 1. "Untagged" todos (inbox)
- [x] 2. Todos due today
- [X] 3: starred todos
- [ ] 4. Recurring todos ("routines")
- [x] 5. All todos
- [X] 6. Todos with a specific tag

Notice that the "task" kind in elos is represented by the number 17 (an enum).
The filter op is also an enumeration, 2 is equality, 4 is less than equal, 5 is greater than equal, and 7 is contained in. These are all found in data.proto in the elos/x/data directory. The 'Reference' object is intentionally capitalized to not collide with the native protobuf json of a oneof type. The kind indicates the primitive's type, and the value it's value. This is polymorphic.

These are the query objects in JSON for the various queries:
1. len(tags) == 0:
```
{'kind': 17, 'filters': [{ 'op':1, 'fns': [2], 'field': 'tags', 'Reference': { 'kind': 3, 'value': 0 }}]}
```
2. between two time instants (specified in seconds since epoch)
```
{'kind': 17, 'filters':[{'op': 4, 'field': 'deadline_at.seconds', 'Reference': {'kind': 3, 'value': ENDTIME_INT}}, {'op': 5, 'field': 'deadline_at.seconds', 'Reference': {'kind': 3, 'value':STARTTIME_INT}}]}
```
3. tags list contains "starred"
```
{'kind': 17, 'filters': [{'op': 7, 'field': 'tags', 'Reference': {'kind': 1, 'value': 'starred'}}]}
```
4. 
5. all todos
```
{'kind': 17}
```
6. Todos with tag name, TAGNAME
```
{'kind': 17, 'filters': [{'op': 7, 'field': 'tags', 'Reference': {'kind': 1, 'value': 'TAGNAME'}}]}
```

### CODE REVIEW

**Comment by nlandolfi** [comment]
@viggyfresh I don't know how we want to do routines?

File: data/data.proto [hunk 1]
```diff
@@ -113,6 +113,31 @@ message Filter {
...
+
+        // NI corresponds to the set exclusion relation.
+        //
+        // The field MUST be an array, or map.
+        NI = 8;
+    }
+
+    // EXPERIMENTAL
+    enum Fn {
+        // UNKNOWN corresponds to undefined behavior.
```
Comment (nlandolfi) [inline data/data.proto]: should be NOP not UNKNOWN

File: data/data.proto [hunk 2]
```diff
@@ -113,6 +113,31 @@ message Filter {
...
+        //
+        // In any case, UNKNOWN may be treated as a no-op.
+        NOP = 0;
+
+        // True if nil
```
Comment (nlandolfi) [inline data/data.proto]: field is nil, so the return type is boolean; make this clear in documentaion

File: data/filter.go [hunk 3]
```diff
@@ -2,44 +2,126 @@ package data
 
 import (
 	"encoding/json"
+	"log"
```
Comment (nlandolfi) [inline data/filter.go]: should be no logging

File: data/filter.go [hunk 4]
```diff
@@ -2,44 +2,126 @@ package data
...
-		if p.OrigName == f.Field {
-			fieldName = p.Name
+	v := reflect.ValueOf(rec)
+	var ok bool
+
+	path := strings.Split(f.Field, ".")
+	for _, p := range path {
+		v, ok = fieldFrom(v, p)
+		if !ok {
+			log.Printf("failed to retrieve field: %v, %v", rec, p)
```
Comment (nlandolfi) [inline data/filter.go]: remove

File: data/filter.go [hunk 5]
```diff
@@ -2,44 +2,126 @@ package data
...
+			return false
 		}
 	}
 
-	if fieldName == "" {
-		return false
+	for i := len(f.Fns) - 1; i >= 0; i-- {
```
Comment (nlandolfi) [inline data/filter.go]: document somewhere function application order!!

File: data/filter.go [hunk 6]
```diff
@@ -2,44 +2,126 @@ package data
...
+		return !referenceContains(f.Reference, v)
 	case Filter_UNKNOWN:
 		fallthrough
 	default:
 		return true // always match if we don't know how to filter
 	}
 }
 
+func apply(v reflect.Value, fn Filter_Fn) reflect.Value {
+	switch fn {
```
Comment (nlandolfi) [inline data/filter.go]: remove this comment

File: data/data.proto [hunk 7]
```diff
@@ -122,21 +122,20 @@ message Filter {
...
+        // True if nil, returns a boolean.
         NIL = 1;
 
-        // Number of items in list, Number of key/val pairs in map
+        // Number of items in list, Number of key/val pairs in map.
+        // Returns an integer.
         LEN = 2;
 
-        // Lower cases a string
+        // Lower cases a string, Returns a string.
```
Comment (viggyfresh) [inline data/data.proto]: nit - returns doesn't need to be capitalized

File: data/data.proto [hunk 8]
```diff
@@ -122,21 +122,20 @@ message Filter {
...
         DNC = 3;
 
-        // Upper cases a string
+        // Upper cases a string, Returns a string.
```
Comment (viggyfresh) [inline data/data.proto]: same as above

File: data/data.proto [hunk 9]
```diff
@@ -122,21 +122,20 @@ message Filter {
...
+        // NOP corresponds the identity function. Returns the same value.
         NOP = 0;
 
-        // True if nil
+        // True if nil, returns a boolean.
         NIL = 1;
 
-        // Number of items in list, Number of key/val pairs in map
+        // Number of items in list, Number of key/val pairs in map.
+        // Returns an integer.
```
Comment (viggyfresh) [inline data/data.proto]: this is almost Pythonic haha - love it

**Comment by viggyfresh** [comment]
Ooh - didnt' realize your comment had been edited with examples. Thankee good sir

**Comment by viggyfresh** [comment]
Thinking big picture, the one concern I have is keeping the constant values within enums in synch. Any bright ideas for sharing those concerns?

**Comment by nlandolfi** [comment]
For sample queries in JSON, see the PR description, or if that's not clear let me know.

For syncing the enum values, I think it is definitely debt at the moment. Ideally, we would just generate the protobuf files for javascript as well. Which is something that I think we could get right. 

The good guarantee about protobufs is so long as I don't do anything stupid (i.e., manually change the enum values in protobuf file, they will always be backwards compatible). AKA the constants in the app can be an append only list that never changes.

Additionally I think there will be a benefit for some sort of DSL in javascript for some of the more tedious things, such as, e.g., 

```
query(Tasks, equality("name", stringReference("foobar"))
```

**Comment by viggyfresh** [comment]
I agree - just wanted to flag the temporary debt as I see it.

For now, I've got a separate JS file managing all Elos constants, should be easy to rip it out eventually haha

**Comment by viggyfresh** [comment]
I think I'm assembling these correctly, but to be sure, can you update with latest models?

**Comment by nlandolfi** [comment]
I think the latest models are on master, and that there have been no relevant changes to the task model since I pushed this.

**Comment by nlandolfi** [comment]
For posterity, the enum for filter ref kinds is not in the protos, and is defined [here](https://github.com/elos/x/blob/master/data/filter.go#L354)

**Pull request merged**